resolwe/flow/execution_engines/workflow/__init__.py

```diff
@@ -0,0 +1,106 @@
+"""An execution engine that supports workflow specifications."""
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import collections
+
+from six import string_types
+
+from django.db import transaction
+
+from resolwe.flow.execution_engines.base import BaseExecutionEngine
+from resolwe.flow.execution_engines.exceptions import ExecutionError
+from resolwe.flow.expression_engines import EvaluationError
+from resolwe.flow.models import Data, Process
+
+
+class ExecutionEngine(BaseExecutionEngine):
+    """An execution engine that supports workflow specifications."""
+
+    name = 'workflow'
+
+    def get_output_schema(self, process):
+        """Return any additional output schema for the process."""
+        return [
+            {'name': 'steps', 'type': 'list:data:'},
+        ]
+
+    @transaction.atomic
+    def evaluate(self, data):
+        """Evaluate the code needed to compute a given Data object."""
+        expression_engine = data.process.requirements.get('expression-engine', None)
+        if expression_engine is not None:
+            expression_engine = self.get_expression_engine(expression_engine)
+
+        # Parse steps.
+        steps = data.process.run.get('program', None)
+        if steps is None:
+            return
+
+        if not isinstance(steps, list):
+            raise ExecutionError('Workflow program must be a list of steps.')
```

Comment:
The problem with the monolithic schema is that it makes everything non-modular as schema for all supported execution engines must be defined in one place.

Because the definition of the program depends on the execution engine and only the execution engine defines the schema. This is why I updated the schema to make the program be of any type (string, object, array).